docs/help/_posts/2010-04-01-argument-matchers.markdown

```diff
@@ -156,10 +165,33 @@ subject.StartWithWidget(4);
 widgetFactory.Received().Make(Arg.Is<int>(x => x > 0));
 ```
 
-In this example it would be an error to use an argument matcher in the `ACT` part of this test. Even if we don't mind what specific argument we pass to our subject, `Arg.Any` is only for substitutes, and only for setting return values or checking received calls; not for real calls.
+In this example it would be an error to use an argument matcher in the `ACT` part of this test. Even if we don't mind what specific argument we pass to our subject, `Arg.Any` is only for substitutes, and only for specifying a call while setting return values, checking received calls or for configuring callbacks; not for real calls.
 
 (If you do want to indicate to readers that the precise argument used for a real call doesn't matter you could use a variable such as `var someInt = 4; subject.StartWithWidget(someInt);` or similar. Just stay clear of argument matchers for this!)
 
+We can also use argument matchers for configuring callbacks:
+
+```csharp
+/* ARRANGE */
+var widgetFactory = Substitute.For<IWidgetFactory>();
+var subject = new Sprocket(widgetFactory);
+
+// OK: Use arg matcher to configure a callback:
+var makeCalled = false;
+widgetFactory.When(x => x.Make(Arg.Any<int>())).Do(x => makeCalled = true);
```

Comment:
The code snippet is very nice, but I'm afraid that we might demonstrate people anti-pattern here, as `Received()` should be used for this particular case. They might not remember later the exact context, but will refer like "I saw somewhere in the official docs them doing this" (happens to me often 😟 ).

Instead, we might use `Do` here for the exactly same purpose as we use `Arg.Do()` below:

```c#
...Do(x => makeCalledForId_Way1 = x.ArgAt<int>(0));
```

Reply:
Very good point, but I think we'd need much larger changes to the docs to get better examples. Even for this we should really do `.Received().Make(42)` rather than catching `42` and asserting.

Reply:
Yes, your concern about `Received()` is fully valid.

I got another idea how we could prettify it. What if we change interface to the following:
```c#
public class WidgetInfo {
    public string Name { get; set; }
}

public interface IWidgetFactory {
  object Make(WidgetInfo info);
}
```

Then we can improve our samples as following:
```c#
// NOT OK: using an arg matcher as a value, not to specify a call:
//    widgetFactory.Make(Arg.Any<WidgetInfo>()).Returns(Arg.Any<object>());

// Instead use something like:
widgetFactory.Make(Arg.Any<WidgetInfo>()).Returns(new { Name = "Any widget" });
```

and

```c#
var widgetFactory = Substitute.For<IWidgetFactory>();
var subject = new Sprocket(widgetFactory);

// OK: Use arg matcher to configure a callback:
var log = new List<string>();
widgetFactory.When(x => x.Make(Arg.Any<WidgetInfo>()).Do(x => log.Add(x.Arg<WidgetInfo>().Name));

// OK: Use Arg.Do to configure a callback:
var log2 = new List<string>();
widgetFactory.Make(Arg.Do<WidgetInfo>(info => log2.Add(info.Name)));
```

It would be a _perfectly_ valid scenario, when you would indeed use callbacks instead of asserting that via ugly "Arg.Is()` inspection.